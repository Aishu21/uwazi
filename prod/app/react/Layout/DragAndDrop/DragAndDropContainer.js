"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.default = exports.containerTarget = exports.DragAndDropContainer = void 0;var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireWildcard(require("react"));
var _reactDnd = require("react-dnd");
var _reactDndHtml5Backend = _interopRequireDefault(require("react-dnd-html5-backend"));

var _uniqueID = _interopRequireDefault(require("../../../shared/uniqueID"));
var _DragAndDropItem = _interopRequireDefault(require("./DragAndDropItem"));function _interopRequireWildcard(obj) {if (obj && obj.__esModule) {return obj;} else {var newObj = {};if (obj != null) {for (var key in obj) {if (Object.prototype.hasOwnProperty.call(obj, key)) {var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};if (desc.get || desc.set) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}}newObj.default = obj;return newObj;}}function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}var REACT_ELEMENT_TYPE;function _jsx(type, props, key, children) {if (!REACT_ELEMENT_TYPE) {REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol["for"] && Symbol["for"]("react.element") || 0xeac7;}var defaultProps = type && type.defaultProps;var childrenLength = arguments.length - 3;if (!props && childrenLength !== 0) {props = { children: void 0 };}if (props && defaultProps) {for (var propName in defaultProps) {if (props[propName] === void 0) {props[propName] = defaultProps[propName];}}} else if (!props) {props = defaultProps || {};}if (childrenLength === 1) {props.children = children;} else if (childrenLength > 1) {var childArray = new Array(childrenLength);for (var i = 0; i < childrenLength; i++) {childArray[i] = arguments[i + 3];}props.children = childArray;}return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null };}

class DragAndDropContainer extends _react.Component {
  constructor(props) {
    super(props);
    this.state = { id: props.id || (0, _uniqueID.default)() };
    this.container = { id: this.state.id };
    this.moveItem = this.moveItem.bind(this);
    this.removeItem = this.removeItem.bind(this);
    this.renderItem = this.renderItem.bind(this);
  }

  moveItem(_dragIndex, hoverIndex, item) {
    const dragIndex = _dragIndex;
    const items = this.props.items.concat();
    if (!items.find(_item => _item.id === item.id)) {
      return;
    }
    items.splice(dragIndex, 1);
    items.splice(hoverIndex, 0, item.originalItem || item);
    this.props.onChange(items);
  }

  removeItem(id) {
    let items = this.props.items.concat();
    items = items.filter(item => item.id !== id);
    this.props.onChange(items);
  }

  renderItem(item, index) {
    if (this.props.renderItem) {
      return this.props.renderItem(item, index);
    }

    return item.content;
  }

  render() {
    const { connectDropTarget, items, iconHandle } = this.props;
    return (
      _jsx("div", {}, void 0,
      connectDropTarget(
      _jsx("ul", { className: "list-group" }, void 0,
      items.map((item, index) =>
      _jsx(_DragAndDropItem.default, {
        moveItem: this.moveItem,
        removeItem: this.removeItem,
        index: index,
        iconHandle: iconHandle || !!item.items,

        name: item.name,
        container: this.container,
        items: item.items,
        id: item.id,
        originalItem: item }, item.id,

      this.renderItem)),


      _jsx("div", { className: "no-properties" }, void 0,
      _jsx("div", { className: "no-properties-wrap" }, void 0, "Drag items here"))))));




  }}exports.DragAndDropContainer = DragAndDropContainer;


DragAndDropContainer.defaultProps = {
  iconHandle: false };












const containerTarget = {
  drop(props, monitor, component) {
    const item = monitor.getItem();
    if (item.id === component.state.id) {
      return;
    }
    if (!monitor.getDropResult() || !monitor.getDropResult().id) {
      const items = props.items.concat();
      if (!items.find(_item => _item.id === item.id)) {
        items.push(item.originalItem || item);
        props.onChange(items);
      }
      return { id: component.state.id };
    }
  } };exports.containerTarget = containerTarget;


const dragAndDropContainer = (0, _reactDnd.DropTarget)('DRAG_AND_DROP_ITEM', containerTarget, connect => ({
  connectDropTarget: connect.dropTarget() }))(
DragAndDropContainer);var _default =

(0, _reactDnd.DragDropContext)(_reactDndHtml5Backend.default)(dragAndDropContainer);exports.default = _default;