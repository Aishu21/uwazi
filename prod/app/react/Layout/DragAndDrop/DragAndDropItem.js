"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.default = exports.DragAndDropItem = exports.itemTarget = exports.itemSource = void 0;var _reactDnd = require("react-dnd");
var _reactDom = require("react-dom");
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireWildcard(require("react"));
var _reactRedux = require("react-redux");

var _UI = require("../../UI");function _interopRequireWildcard(obj) {if (obj && obj.__esModule) {return obj;} else {var newObj = {};if (obj != null) {for (var key in obj) {if (Object.prototype.hasOwnProperty.call(obj, key)) {var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};if (desc.get || desc.set) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}}newObj.default = obj;return newObj;}}function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}var REACT_ELEMENT_TYPE;function _jsx(type, props, key, children) {if (!REACT_ELEMENT_TYPE) {REACT_ELEMENT_TYPE = typeof Symbol === "function" && Symbol["for"] && Symbol["for"]("react.element") || 0xeac7;}var defaultProps = type && type.defaultProps;var childrenLength = arguments.length - 3;if (!props && childrenLength !== 0) {props = { children: void 0 };}if (props && defaultProps) {for (var propName in defaultProps) {if (props[propName] === void 0) {props[propName] = defaultProps[propName];}}} else if (!props) {props = defaultProps || {};}if (childrenLength === 1) {props.children = children;} else if (childrenLength > 1) {var childArray = new Array(childrenLength);for (var i = 0; i < childrenLength; i++) {childArray[i] = arguments[i + 3];}props.children = childArray;}return { $$typeof: REACT_ELEMENT_TYPE, type: type, key: key === undefined ? null : '' + key, ref: null, props: props, _owner: null };}

const itemSource = {
  beginDrag(props) {
    return {
      id: props.id,
      name: props.name,
      items: props.items,
      container: props.container,
      index: props.index,
      originalItem: props.originalItem };

  },

  endDrag(props, monitor) {
    const container = monitor.getDropResult();
    if (container && container.id !== props.container.id) {
      props.removeItem(props.id);
    }
  } };exports.itemSource = itemSource;


const itemTarget = {
  hover(props, monitor, component) {
    const item = monitor.getItem();
    const dragIndex = item.index;
    const hoverIndex = props.index;

    if (props.id === item.id) {
      return;
    }

    if (props.container.id !== item.container.id) {
      return;
    }

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = (0, _reactDom.findDOMNode)(component).getBoundingClientRect(); // eslint-disable-line react/no-find-dom-node

    // Get vertical middle
    const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    // Get pixels to the top
    const hoverClientY = clientOffset.y - hoverBoundingRect.top;

    // Only perform the move when the mouse has crossed half of the items height
    // When dragging downwards, only move when the cursor is below 50%
    // When dragging upwards, only move when the cursor is above 50%

    // Dragging downwards
    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
      return;
    }

    // Dragging upwards
    if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
      return;
    }

    // Time to actually perform the action
    props.moveItem(dragIndex, hoverIndex, item);

    // Note: we're mutating the monitor item here!
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    item.index = hoverIndex;
    return null;
  } };exports.itemTarget = itemTarget;


class DragAndDropItem extends _react.Component {
  render() {
    const { iconHandle, isDragging, connectDragPreview, connectDragSource, connectDropTarget } = this.props;
    let propertyClass = 'list-group-item';
    if (isDragging) {
      propertyClass += ' dragging';
    }

    if (!iconHandle) {
      propertyClass += ' draggable';
    }

    const result = connectDropTarget(
    _jsx("div", { className: propertyClass }, void 0,
    iconHandle ? connectDragSource(_jsx("span", { className: "draggable" }, void 0, _jsx(_UI.Icon, { icon: "bars" }))) : _jsx(_UI.Icon, { icon: "bars" }),
    this.props.children(this.props.originalItem, this.props.index)));



    if (!iconHandle) {
      return connectDragSource(result);
    }

    return connectDragPreview(result);
  }}exports.DragAndDropItem = DragAndDropItem;


DragAndDropItem.defaultProps = {
  iconHandle: false,
  children: () => {} };















let dragAndDropItem = (0, _reactDnd.DropTarget)('DRAG_AND_DROP_ITEM', itemTarget, connector => ({
  connectDropTarget: connector.dropTarget() }))(
DragAndDropItem);

dragAndDropItem = (0, _reactDnd.DragSource)('DRAG_AND_DROP_ITEM', itemSource, (connector, monitor) => ({
  connectDragSource: connector.dragSource(),
  connectDragPreview: connector.dragPreview(),
  isDragging: monitor.isDragging() }))(
dragAndDropItem);var _default =

(0, _reactRedux.connect)()(dragAndDropItem);exports.default = _default;