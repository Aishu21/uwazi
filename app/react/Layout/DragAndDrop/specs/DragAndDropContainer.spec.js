import React from 'react';
import { shallow } from 'enzyme';
import { DragAndDropContainer, containerTarget } from '../DragAndDropContainer';
import DragAndDropItem from '../DragAndDropItem';

describe('DragAndDropContainer', () => {
  describe('react component', () => {
    let component;
    let props;
    let items;
    beforeEach(() => {
      items = [
        { content: 'A rude awakening', id: 1 },
        { content: 'A nice awakening', id: 2 }
      ];
      props = {
        connectDropTarget: jasmine.createSpy('connectDropTarget').and.callFake(value => value),
        items,
        onChange: jasmine.createSpy('onChange')
      };
    });

    const render = () => {
      component = shallow(<DragAndDropContainer {...props} />);
    };

    it('should render a DragAndDropItem for each item', () => {
      render();
      expect(component.find(DragAndDropItem).length).toBe(2);
      expect(component.find(DragAndDropItem).first().props().children).toBe('A rude awakening');
    });

    describe('accepts a custom render function', () => {
      it('to render items', () => {
        props.renderItem = () => <span>Avocado</span>;
        render();
        expect(component.find(DragAndDropItem).first().find('span').text()).toBe('Avocado');
      });
    });

    describe('when an item moves', () => {
      it('should reorder the items and call onChange', () => {
        render();
        component.instance().moveItem(1, 0, items[1]);
        expect(props.onChange).toHaveBeenCalledWith([{ content: 'A nice awakening', id: 2 }, { content: 'A rude awakening', id: 1 }]);
      });

      describe('when the item does not belong to the container', () => {
        it('should do nothing', () => {
          render();
          component.instance().moveItem(1, 0, { id: 27 });
          expect(props.onChange).not.toHaveBeenCalled();
        });
      });
    });

    describe('removeItem', () => {
      it('should remove the item and call onChange', () => {
        render();
        component.instance().removeItem(1);
        expect(props.onChange).toHaveBeenCalledWith([{ content: 'A nice awakening', id: 2 }]);
      });
    });
  });

  describe('containerTarget', () => {
    describe('drop', () => {
      it('should return the container id', () => {
        const monitor = {
          getItem: () => ({ id: 1 }),
          getDropResult: () => {}
        };

        const component = { state: { id: 'container_id' } };
        const props = {
          items: [{ id: 1 }],
          onChange: jasmine.createSpy('onChange')
        };
        const dropResult = containerTarget.drop(props, monitor, component);
        expect(dropResult).toEqual({ id: 'container_id' });
        expect(props.onChange).not.toHaveBeenCalled();
      });

      it('should add the item if missing', () => {
        const monitor = {
          getItem: () => ({ id: 1 }),
          getDropResult: () => {}
        };
        const component = { state: { id: 'container_id' } };
        const props = {
          items: [],
          onChange: jasmine.createSpy('onChange')
        };
        containerTarget.drop(props, monitor, component);
        expect(props.onChange).toHaveBeenCalledWith([{ id: 1 }]);
      });

      it('should NOT add the item if other container handles the drop', () => {
        const monitor = {
          getItem: () => ({ id: 1 }),
          getDropResult: () => ({ id: 'I_handle_this' })
        };
        const component = { state: { id: 'container_id' } };
        const props = {
          items: [],
          onChange: jasmine.createSpy('onChange')
        };
        containerTarget.drop(props, monitor, component);
        expect(props.onChange).not.toHaveBeenCalled();
      });
    });
  });
});
