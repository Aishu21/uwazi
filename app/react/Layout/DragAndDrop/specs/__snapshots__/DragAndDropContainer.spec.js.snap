// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragAndDropContainer react component should enable iconHandle on children if iconHandle is enabled in container 1`] = `
<div>
  <ul
    className="list-group"
  >
    <Connect(DragSource(DropTarget(DragAndDropItem)))
      container={
        Object {
          "id": "opr5tpw56w",
        }
      }
      iconHandle={true}
      id={1}
      index={0}
      key="1"
      moveItem={[Function]}
      originalItem={
        Object {
          "content": "A rude awakening",
          "id": 1,
        }
      }
      removeItem={[Function]}
    />
    <Connect(DragSource(DropTarget(DragAndDropItem)))
      container={
        Object {
          "id": "opr5tpw56w",
        }
      }
      iconHandle={true}
      id={2}
      index={1}
      key="2"
      moveItem={[Function]}
      originalItem={
        Object {
          "content": "A nice awakening",
          "id": 2,
        }
      }
      removeItem={[Function]}
    />
    <div
      className="no-properties"
    >
      <div
        className="no-properties-wrap"
      >
        Drag items here
      </div>
    </div>
  </ul>
</div>
`;
